src/ILLink.RoslynAnalyzer/COMAnalyzer.cs

```diff
@@ -0,0 +1,124 @@
+				if (!context.Options.IsMSBuildPropertyValueTrue (MSBuildPropertyOptionNames.EnableTrimAnalyzer, compilation))
+					return;
+
+				context.RegisterOperationAction (operationContext => {
+					var invocationOperation = (IInvocationOperation) operationContext.Operation;
+					var targetMethod = invocationOperation.TargetMethod;
+					if (!targetMethod.HasAttribute (DllImportAttribute))
+						return;
+
+					foreach (var parameter in targetMethod.Parameters) {
+						if (IsComInterop (parameter)) {
+							operationContext.ReportDiagnostic (Diagnostic.Create (s_correctnessOfCOMCannotBeGuaranteed,
+								operationContext.Operation.Syntax.GetLocation (), targetMethod.GetDisplayName ()));
+						}
+					}
+
+					if (IsComInterop (targetMethod.ReturnType)) {
+						operationContext.ReportDiagnostic (Diagnostic.Create (s_correctnessOfCOMCannotBeGuaranteed,
+								operationContext.Operation.Syntax.GetLocation (), targetMethod.GetDisplayName ()));
+					}
+				}, OperationKind.Invocation);
+			});
+
+			static bool IsComInterop (ISymbol symbol)
+			{
+				if (symbol.TryGetAttribute (MarshalAsAttribute, out var marshalAsAttribute) &&
+					marshalAsAttribute.ConstructorArguments.Length >= 1 && marshalAsAttribute.ConstructorArguments[0] is TypedConstant typedConstant &&
+					typedConstant.Type != null && typedConstant.Type.IsUnmanagedType) {
+					var unmanagedType = typedConstant.Value;
+					switch (unmanagedType) {
+					case (int) UnmanagedType.IUnknown:
+					case (int) UnmanagedType.IDispatch:
+					case (int) UnmanagedType.Interface:
+						return true;
+
+					default:
+						break;
+					}
+				}
+
+				if (IsInterface (symbol))
+					return true;
+
+				if (symbol is not IParameterSymbol parameterSymbol)
+					return false;
+
+				var parameterType = parameterSymbol.Type;
+				if (parameterType.ContainingNamespace.Name == "System" && parameterType.Name == "Array") {
+					// System.Array marshals as IUnknown by default
+					return true;
+				} else if (parameterType.ContainingNamespace.Name == "System" && parameterType.Name == "String" ||
+					parameterType.ContainingNamespace.Name == "System.Text" && parameterType.Name == "StringBuilder") {
+					// String and StringBuilder are special cased by interop
+					return false;
+				}
+
+				if (parameterType.IsValueType) {
+					// Value types don't marshal as COM
+					return false;
+				} else if (IsInterface (parameterSymbol.Type)) {
+					// Interface types marshal as COM by default
+					return true;
+				} else if (parameterType.ContainingNamespace.Name == "System" &&
+					parameterType.Name == "MulticastDelegate") {
+					// Delegates are special cased by interop
+					return false;
+				} else if (parameterType.ContainingNamespace.Name == "System.Runtime.InteropServices" &&
+					(parameterType.Name == "CriticalHandle" || parameterType.Name == "SafeHandle")) {
+					// Subclasses of CriticalHandle and SafeHandle are special cased by interop
```

Comment:
Do we need to check for subclasses here?

Reply:
Yup. Added a helper method in ISymbolExtensions for this. I moved the `IsInterface` method there as well.